sbt-bridge/src/dotty/tools/xsbt/CompilerBridgeDriver.java

```diff
@@ -50,15 +55,68 @@ public boolean sourcesRequired() {
     return false;
   }
 
+  private static VirtualFile asVirtualFile(SourceFile sourceFile, DelegatingReporter reporter,
+      HashMap<AbstractFile, VirtualFile> lookup, Map<AbstractFile, VirtualFile> placeholders) {
+    VirtualFile maybeCached = lookup.get(sourceFile.file());
+    if (maybeCached != null) {
+      return maybeCached;
+    } else {
+      return fallbackVirtualFile(reporter, sourceFile, placeholders);
+    }
+  }
+
+  private static void reportMissingFile(DelegatingReporter reporter, SourceFile sourceFile) {
+    String underline = String.join("", Collections.nCopies(sourceFile.path().length(), "^"));
+    String message =
+      sourceFile.path() + ": Missing virtual file\n" +
+      underline + "\n" +
+      "    Falling back to placeholder for the given source file (of class " + sourceFile.getClass().getName() + ")\n" +
+      "    This is likely a bug in incremental compilation for the Scala 3 compiler. Please report it to the Scala 3 maintainers.";
+    reporter.reportBasicWarning(message);
+  }
+
+  private static VirtualFile fallbackVirtualFile(DelegatingReporter reporter, SourceFile sourceFile,
+      Map<AbstractFile, VirtualFile> placeholders) {
+    return placeholders.computeIfAbsent(sourceFile.file(), path -> {
+      reportMissingFile(reporter, sourceFile);
+      if (sourceFile.file().jpath() != null)
+        return new BasicPathBasedFile(sourceFile);
+      else
+        return new PlaceholderVirtualFile(sourceFile);
+    });
+  }
+
   synchronized public void run(VirtualFile[] sources, AnalysisCallback callback, Logger log, Reporter delegate) {
-    DelegatingReporter reporter = new DelegatingReporter(delegate);
+    VirtualFile[] sortedSources = new VirtualFile[sources.length];
+    System.arraycopy(sources, 0, sortedSources, 0, sources.length);
+    Arrays.sort(sortedSources, (x0, x1) -> x0.id().compareTo(x1.id()));
+
+    ListBuffer<AbstractFile> sourcesBuffer = new ListBuffer<>();
+    HashMap<AbstractFile, VirtualFile> lookup = new HashMap<>(sources.length, 0.25f);
+
+    for (int i = 0; i < sources.length; i++) {
+      VirtualFile source = sortedSources[i];
+      AbstractFile abstractFile = asDottyFile(source);
+      sourcesBuffer.append(abstractFile);
+      lookup.put(abstractFile, source);
+    }
+
+    HashMap<AbstractFile, VirtualFile> placeholders = new HashMap<>();
```

Comment:
I don't think you need placeholders anymore, you can add entry to lookup instead

Reply:
makes sense :)